org.opent2t.sample.multisensor.superpopular/com.wink.sensorpod/js/thingTranslator.js

```diff
@@ -0,0 +1,181 @@
+    } else if (typeof arg !== expectedType) {
+        throw new Error('Invalid argument: ' + argName + '. ' +
+            'Expected type: ' + expectedType + ', got: ' + (typeof arg));
+    }
+}
+
+// Wink does not always populate every desired_state property, but last_reading doesn't necessarily
+// update as soon as we send our PUT request. Instead of relying just on one state or the other,
+// we use this StateReader class to read from desired_state if it is there, and fall back to last_reading
+// if it is not.
+class StateReader {
+    constructor(desired_state, last_reading) {
+        this.desired_state = desired_state;
+        this.last_reading = last_reading;
+    }
+
+    get(state) {
+        if (this.desired_state[state] !== undefined) {
+            return this.desired_state[state];
+        }
+        else {
+            return this.last_reading[state];
+        }
+    }
+
+    containsKey(state) {
+        return (state in this.desired_state ||
+                state in this.last_reading); 
+    }
+}
+
+// Helper method to convert Wink's 0.0-1.0 humidity scale to a 0-100 scale
+function scaleDeviceHumidityToTranslatorHumidity(humidity) {
+    return Math.floor(humidity * 100);
+}
+
+function createSensor(key, value, resourceType) {
+    return {
+        id: key,
+        rt: resourceType || 'oic.r.sensor',
+        value: value
+    };
+}
+
+function createMotionSensor(key, value) {
+    return createSensor(key, value, 'oic.r.sensor.motion');
+}
+
+function createContactSensor(key, value) {
+    return createSensor(key, value, 'oic.r.sensor.contact');
+}
+
+function createTemperature(key, value) {
+    return {
+        id: key,
+        rt: 'oic.r.temperature',
+        temperature: value,
+        units: "C"
+    };
+}
+
+function createHumidity(key, value) {
+    return {
+        id: key,
+        rt: 'oic.r.humidity',
+        humidity: scaleDeviceHumidityToTranslatorHumidity(value)
+    };
+}
+
+function readSensorsArray(stateReader) {
+    var sensors = [];
+
+    var winkSensors = {
+        'motion': createMotionSensor,
+        'opened': createContactSensor,
+        'temperature': createTemperature,
+        'humidity': createHumidity,
+        'brightness': createSensor,
+        'loudness': createSensor,
+        'vibration': createSensor,
+        'locked': createSensor,
+        'liquid_detected': createSensor,
+        'occupied': createSensor
+    };
+
+    Object.keys(winkSensors).forEach(key => {
+        if (stateReader.containsKey(key)) {
+            var sensor = winkSensors[key](key, stateReader.get(key));
+            sensors.push(sensor);
+        }
+    });
+
+    return sensors;
+}
+
+// Helper method to convert the device schema to the translator schema.
+function deviceSchemaToTranslatorSchema(deviceSchema) {
+
+    var stateReader = new StateReader(deviceSchema.desired_state, deviceSchema.last_reading);
+
+    return {
+        id: deviceSchema['object_type'] + '.' + deviceSchema['object_id'],
```

Comment:
If nothing else it's consistent with the other devices right now, which makes as much sense to me as anything else until we have a plan going forward for these IDs, but if you want me to change it to something else I can.

---
In reply to: [79105944](https://github.com/openT2T/translators/pull/97#discussion_r79105944) [](ancestors = 79105944)